programs/redstone-sol/src/util.rs

```diff
@@ -0,0 +1,159 @@
+        num = (num << 8) | c as u128;
+    }
+    num.to_string()
+}
+
+#[cfg(feature = "dev")]
+pub fn bytes_to_hex(bytes: &[u8]) -> String {
+    use std::fmt::Write;
+    bytes.iter().fold(
+        String::with_capacity(bytes.len() * 2),
+        |mut output, b| {
+            let _ = write!(output, "{:02x}", b);
+            output
+        },
+    )
+}
+
+pub trait FromBytesRepr<T: Sized>: Sized {
+    fn from_bytes(bytes: T) -> Self;
+}
+
+impl FromBytesRepr<&[u8]> for U256 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        let mut array = [0u8; 32];
+        let len = bytes.len().min(32);
+        array[..len].copy_from_slice(&bytes[..len]);
+        array
+    }
+}
+
+impl FromBytesRepr<&[u8]> for usize {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0usize, |acc, &b| (acc << 8) | b as usize)
+    }
+}
+
+impl FromBytesRepr<&[u8]> for u64 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0u64, |acc, &b| (acc << 8) | b as u64)
+    }
+}
+
+impl FromBytesRepr<&[u8]> for u128 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0u128, |acc, &b| (acc << 8) | b as u128)
+    }
+}
+
+pub fn calculate_median(values: &mut [U256]) -> Option<U256> {
+    match values.len() {
+        0 => None,
+        1 => Some(values[0]),
+        2 => Some(avg_u256(&values[0], &values[1])),
+        3 => {
+            let (a, b, c) = (&values[0], &values[1], &values[2]);
+            if (b >= a && b <= c) || (b >= c && b <= a) {
+                Some(*b)
+            } else if (a >= b && a <= c) || (a >= c && a <= b) {
+                Some(*a)
+            } else {
+                Some(*c)
+            }
+        }
+        _ => {
+            values.sort_unstable();
+            let len = values.len();
+
+            if len % 2 == 0 {
+                let mid = len / 2;
+                let left = &values[mid - 1];
+                let right = &values[mid];
+                Some(avg_u256(left, right))
+            } else {
+                Some(values[len / 2])
+            }
+        }
+    }
+}
+
+fn avg_u256(a: &U256, b: &U256) -> U256 {
+    divide_u256_by_2(&add_u256(a, b))
+}
+
+fn add_u256(a: &U256, b: &U256) -> U256 {
+    let mut result = [0u8; 32];
+    let mut carry = 0u16;
+
+    for i in (0..32).rev() {
+        let sum = u16::from(a[i]) + u16::from(b[i]) + carry;
+        result[i] = sum as u8;
+        carry = sum >> 8;
+    }
+
+    if carry > 0 {
+        U256_MAX
+    } else {
+        result
+    }
+}
+
+fn divide_u256_by_2(a: &U256) -> U256 {
+    if a == &U256_MAX {
+        return U256_MAX;
+    }
+
+    let mut result = [0u8; 32];
+    let mut carry = 0u8;
+
+    for i in 0..32 {
+        let current = (carry << 7) | (a[i] >> 1);
+        result[i] = current;
+        carry = a[i] & 1;
+    }
+
+    result
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_median_with_max_values() {
+        let mut values = vec![U256_MAX; 100];
```

Comment:
what's the result of median for

```[U256_MAX, U256_MAX, U256_MAX - 10, U256_MAX - 10]``` :>?

Reply:
193444c47cbc9b565f9c64d08fc6e1ade6e87756

Reply:
I used four u64 based u256 from `zkp_u256` crate and convert only before median calculation, CU stays the same

I also added a fuzzy benchmark for median calculation with u128 counterparts, included the overflow case and the case with `[U256_MAX, U256_MAX, U256_MAX - 10, U256_MAX - 10]` 😄